homeassistant/components/input_number/__init__.py

```diff
@@ -80,22 +110,82 @@ def _cv_input_number(cfg):
     extra=vol.ALLOW_EXTRA,
 )
 RELOAD_SERVICE_SCHEMA = vol.Schema({})
+STORAGE_KEY = DOMAIN
+STORAGE_VERSION = 1
+
+
+class NumberStorageCollection(collection.StorageCollection):
+    """Input boolean collection stored in storage."""
+
+    CREATE_SCHEMA = vol.Schema(vol.All(CREATE_FIELDS, _cv_input_number))
+    UPDATE_SCHEMA = vol.Schema(UPDATE_FIELDS)
+
+    async def _process_create_data(self, data: typing.Dict) -> typing.Dict:
+        """Validate the config is valid."""
+        return self.CREATE_SCHEMA(data)
 
+    @callback
+    def _get_suggested_id(self, info: typing.Dict) -> str:
+        """Suggest an ID based on the config."""
+        return info[CONF_NAME]
 
-async def async_setup(hass, config):
+    async def _update_data(self, data: dict, update_data: typing.Dict) -> typing.Dict:
+        """Return a new updated data object."""
+        update_data = self.UPDATE_SCHEMA(update_data)
+        return self.CREATE_SCHEMA({**data, **update_data})
```

Comment:
Is there a way to make this better? For this to work, `CREATE_FIELDS` allow optional `CONF_ID` key.
alternatively the update schema may require both `CONF_MIN` and `CONF_MAX` key present, but then what  if an `UPDATE` modifies `CONF_MIN` or `CONF_MAX` outside of the existing `CONF_INITIAL` value?

Reply:
Update should validate that all values still are ok.  In your example `_update_data` should raise a `ValueError`

Reply:
As per our discussion, `return _cv_input_num({**data, **update_data})` should take care of the update+existing data validation.